## Resilient Distributed Datasets (RDDs) - Lab

Resilient Distributed Datasets (RDD) are fundamental data structures of Spark. An RDD is essentially the Spark representation of a set of data, spread across multiple machines, with APIs to let you act on it. An RDD can come from any data source, e.g. text files, a database, a JSON file, etc.


## Objectives

You will be able to:

- Apply the map(func) transformation to a given function on all elements of an RDD in different partitions 
- Apply a map transformation for all elements of an RDD 
- Compare the difference between a transformation and an action within RDDs 
- Use collect(), count(), and take() actions to trigger spark transformations  
- Use filter to select data that meets certain specifications within an RDD 
- Set number of partitions for parallelizing RDDs 
- Create RDDs from Python collections 


## What are RDDs? 

To get a better understanding of RDDs, let's break down each one of the components of the acronym RDD:

**Resilient:** RDDs are considered "resilient" because they have built-in fault tolerance. This means that even if one of the nodes goes offline, RDDs will be able to restore the data. This is already a huge advantage compared to standard storage. If a standard computer dies while performing an operation, all of its memory will be lost in the process. With RDDs, multiple nodes can go offline, and the action will still be held in working memory.

**Distributed:** The data is contained on multiple nodes of a cluster-computing operation. It is efficiently partitioned to allow for parallelism.

**Dataset:** The dataset has been * partitioned * across the multiple nodes. 

RDDs are the building block upon which more high-level Spark operations are based upon. Chances are, if you are performing an action using Spark, the operation involves RDDs. 

Key Characteristics of RDDs:

- **Immutable:** Once an RDD is created, it cannot be modified. 
- **Lazily Evaluated:** RDDs will not be evaluated until an action is triggered. Essentially, when RDDs are created, they are programmed to perform some action, but that function will not get activated until it is explicitly called. The reason for lazy evaluation is that allows users to organize the actions of their Spark program into smaller actions. It also saves unnecessary computation and memory load.
- **In-Memory:** The operations in Spark are performed in-memory rather than in the database. This is what allows Spark to perform fast operations with very large quantities of data.


### RDD Transformations vs Actions

In Spark, we first create a __base RDD__ and then apply one or more transformations to that base RDD following our processing needs. Being immutable means, **once an RDD is created, it cannot be changed**. As a result, **each transformation of an RDD creates a new RDD**. Finally, we can apply one or more **actions** to the RDDs. Spark uses lazy evaluation, so transformations are not actually executed until an action occurs.


<img src="./images/rdd_diagram.png" width=500>

### Transformations

Transformations create a new dataset from an existing one by passing each dataset element through a function and returning a new RDD representing the results. In short, creating an RDD from an existing RDD is ‘transformation’.
All transformations in Spark are lazy. They do not compute their results right away. Instead, they just remember the transformations applied to some base dataset (e.g. a file). The transformations are only computed when an action requires a result that needs to be returned to the driver program.
A transformation is an RDD that returns another RDD, like map, flatMap, filter, reduceByKey, join, cogroup, etc.

### Actions
Actions return final results of RDD computations. Actions trigger execution using lineage graph to load the data into original RDD and carry out all intermediate transformations and return the final results to the driver program or writes it out to the file system. An action returns a value (to a Spark driver - the user program).

Here are some key transformations and actions that we will explore.


| Transformations   | Actions       |
|-------------------|---------------|
| map(func)         | reduce(func)  |
| filter(func)      | collect()     |
| groupByKey()      | count()       |
| reduceByKey(func) | first()       |
| mapValues(func)   | take()        |
| sample()          | countByKey()  |
| distinct()        | foreach(func) |
| sortByKey()       |               |


Let's see how transformations and actions work through a simple example. In this example, we will perform several actions and transformations on RDDs in order to obtain a better understanding of Spark processing. 

### Create a Python collection 

We need some data to start experimenting with RDDs. Let's create some sample data and see how RDDs handle it. To practice working with RDDs, we're going to use a simple Python list.

- Create a Python list `data` of integers between 1 and 1000 using the `range()` function. 
- Sanity check: confirm the length of the list (it should be 1000)

In [2]:
data = range(1,1001)
len(data)

# 1000

1000

### Initialize an RDD

When using Spark to make computations, datasets are treated as lists of entries. Those lists are split into different partitions across different cores or different computers. Each list of data held in memory is a partition of the RDD. The reason why Spark is able to make computations far faster than other big data processing languages is that it allows all data to be stored __in-memory__, which allows for easy access to the data and, in turn, high-speed processing. Here is an example of how the alphabet might be split into different RDDs and held across a distributed collection of nodes:

<img src ="./images/partitions_1.png" width ="500">

To initialize an RDD, first import `pyspark` and then create a SparkContext assigned to the variable `sc`. Use `'local[*]'` as the master.

In [3]:
import pyspark 
sc = pyspark.SparkContext('local[*]')

22/03/04 22:57:36 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


Once you've created the SparkContext, you can use the `.parallelize()` method to create an RDD that will distribute the list of numbers across multiple cores. Here, create one called `rdd` with 10 partitions using `data` as the collection you are parallelizing.

In [8]:
rdd = sc.parallelize(data, numSlices=10)
print(type(rdd))
# <class 'pyspark.rdd.RDD'>

<class 'pyspark.rdd.PipelinedRDD'>


Determine how many partitions are being used with this RDD with the `.getNumPartitions()` method.

In [9]:
rdd.getNumPartitions()
# 10

10

### Basic descriptive RDD actions

Let's perform some basic operations on our RDD. In the cell below, use the methods:
* `count`: returns the total count of items in the RDD 
* `first`: returns the first item in the RDD
* `take`: returns the first `n` items in the RDD
* `top`: returns the top `n` items
* `collect`: returns everything from your RDD


It's important to note that in a big data context, calling the collect method will often take a very long time to execute and should be handled with care!

In [10]:
# count
rdd.count()

1000

In [11]:
# first
rdd.first()

1

In [12]:
# take
rdd.take(10)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [13]:
# top
rdd.top(10)

[1000, 999, 998, 997, 996, 995, 994, 993, 992, 991]

In [14]:
# collect
## Note: When you are dealing with big data, this could make your computer crash!
## It's best to avoid using the collect() method
rdd.collect()

[1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,
 185

## Map functions

Now that you've been working a little bit with RDDs, let's make this a little more interesting. Imagine you're running a hot new e-commerce startup called BuyStuff, and you're trying to track of how much it charges customers from each item sold. In the next cell, we're going to create simulated data by multiplying the values 1-1000 with a random number from 0-1.

In [15]:
# Run this cell without changes
import numpy as np

np.random.seed(1)
nums = np.array(range(1, 1001))
sales_figures = nums * np.random.rand(1000)
sales_figures

array([4.17022005e-01, 1.44064899e+00, 3.43124452e-04, 1.20933029e+00,
       7.33779454e-01, 5.54031569e-01, 1.30382148e+00, 2.76448582e+00,
       3.57090727e+00, 5.38816734e+00, 4.61113966e+00, 8.22263400e+00,
       2.65787925e+00, 1.22936441e+01, 4.10813898e-01, 1.07274802e+01,
       7.09418164e+00, 1.00564169e+01, 2.66735183e+00, 3.96202978e+00,
       1.68156359e+01, 2.13017547e+01, 7.20875610e+00, 1.66157428e+01,
       2.19097288e+01, 2.32597733e+01, 2.29619371e+00, 1.09353393e+00,
       4.92508217e+00, 2.63442751e+01, 3.04875185e+00, 1.34754440e+01,
       3.16103545e+01, 1.81276197e+01, 2.42156990e+01, 1.13585627e+01,
       2.54005343e+01, 3.17157755e+01, 7.13242816e-01, 3.00057726e+01,
       4.05433046e+01, 3.14229575e+01, 1.20590917e+01, 3.47282905e+01,
       4.64517030e+00, 2.06031022e+01, 4.27039886e+01, 1.40934791e+01,
       1.41009916e+01, 6.50142861e+00, 9.87714851e-01, 3.52994477e+01,
       1.12162901e+01, 1.43395196e+01, 2.70365238e+01, 2.98830253e+00,
      

We now have sales prices for 1000 items currently for sale at BuyStuff. Now create an RDD called `price_items` using the newly created data with 10 slices. After you create it, use one of the basic actions to see what's in the RDD.

In [16]:
price_items = sc.parallelize(sales_figures, numSlices=10)


Now let's perform some operations on this simple dataset. To begin with, create a function that will take into account how much money BuyStuff will receive after sales tax has been applied. Assume a sales tax of 8%, i.e. that the current price data is actually 108% of what it should be.

To make this happen, create a function called `sales_tax()` that returns the amount of money our company will receive after the sales tax has been applied. The function will have this parameter:

* `num`: (float) number to be multiplied by the sales tax.

Apply that function to the RDD by using the `.map()` method and assign it to a variable `renenue_minus_tax`

In [17]:
def sales_tax(num):
    return num/1.08

revenue_minus_tax = price_items.map(sales_tax)

Remember, Spark has __lazy evaluation__, which means that the `sales_tax()` function is a transformer that is not executed until you call an action. Use one of the collection methods to execute the transformer now a part of the RDD and observe the contents of the `revenue_minus_tax` RDD.

In [18]:
# perform action to retrieve rdd values
revenue_minus_tax.collect()

[0.38613148583571666,
 1.3339342471151074,
 0.0003177078259580184,
 1.119750269006814,
 0.6794254204495974,
 0.5129921931599878,
 1.2072421107812004,
 2.559709089207761,
 3.306395618588916,
 4.989043833364416,
 4.269573757811336,
 7.613550004408438,
 2.460999302323821,
 11.38300380506781,
 0.38038323886008557,
 9.932852002642996,
 6.568686703926998,
 9.311497140762526,
 2.4697702160272605,
 3.6685460941644217,
 15.5700332798021,
 19.72384691280254,
 6.674774164502393,
 15.384947014873646,
 20.286785932778663,
 21.536827084351877,
 2.1261052842444474,
 1.012531417148802,
 4.560261266085646,
 24.3928473174837,
 2.8229183785412526,
 12.477262963112656,
 29.26884675459867,
 16.78483304544683,
 22.421943507654227,
 10.517187700202097,
 23.519013263165366,
 29.366458826018675,
 0.6604100152069262,
 27.783122775739535,
 37.54009689367248,
 29.09533100366042,
 11.165825609971687,
 32.15582449246805,
 4.30108360740175,
 19.076946485270035,
 39.540730227199525,
 13.049517705496866,
 13.056473695

### Lambda Functions

Note that you can also use lambda functions if you want to quickly perform simple operations on data without creating a function. Let's assume that BuyStuff has also decided to offer a 10% discount on all of their items on the pre-tax amounts of each item. Use a lambda function within a `.map()` method to apply the additional 10% loss in revenue for BuyStuff and assign the transformed RDD to a new RDD called `discounted`.

In [19]:
discounted = revenue_minus_tax.map(lambda x: 0.9*x)

In [20]:
discounted.take(10)

[0.347518337252145,
 1.2005408224035967,
 0.0002859370433622166,
 1.0077752421061326,
 0.6114828784046377,
 0.461692973843989,
 1.0865178997030804,
 2.303738180286985,
 2.9757560567300243,
 4.490139450027975]

## Chaining Methods

You are also able to chain methods together with Spark. In one line, remove the tax and discount from the revenue of BuyStuff and use a collection method to see the 15 costliest items.

In [22]:
price_items.map(sales_tax).map(lambda x: 0.9*x).take(15)

[0.347518337252145,
 1.2005408224035967,
 0.0002859370433622166,
 1.0077752421061326,
 0.6114828784046377,
 0.461692973843989,
 1.0865178997030804,
 2.303738180286985,
 2.9757560567300243,
 4.490139450027975,
 3.842616382030202,
 6.852195003967594,
 2.214899372091439,
 10.24470342456103,
 0.34234491497407704]

## RDD Lineage


We are able to see the full lineage of all the operations that have been performed on an RDD by using the `RDD.toDebugString()` method. As your transformations become more complex, you are encouraged to call this method to get a better understanding of the dependencies between RDDs. Try calling it on the `discounted` RDD to see what RDDs it is dependent on.

In [23]:
discounted.toDebugString()

b'(10) PythonRDD[11] at RDD at PythonRDD.scala:53 []\n |   ParallelCollectionRDD[6] at readRDDFromFile at PythonRDD.scala:274 []'

### Map vs. Flatmap

Depending on how you want your data to be outputted, you might want to use `.flatMap()` rather than a simple `.map()`. Let's take a look at how it performs operations versus the standard map. Let's say we wanted to maintain the original amount BuyStuff receives for each item as well as the new amount after the tax and discount are applied. Create a map function that will return a tuple with (original price, post-discount price).

In [24]:
mapped = price_items.map(lambda x: (x,(0.9/1.08)*x))
print(mapped.count())
print(mapped.take(10))

1000
[(0.417022004702574, 0.347518337252145), (1.4406489868843162, 1.2005408224035967), (0.0003431244520346599, 0.0002859370433622166), (1.209330290527359, 1.0077752421061326), (0.7337794540855652, 0.6114828784046377), (0.5540315686127868, 0.46169297384398894), (1.3038214796436964, 1.0865178997030802), (2.764485816344382, 2.303738180286985), (3.5709072680760294, 2.9757560567300243), (5.3881673400335695, 4.490139450027974)]


Note that we have 1000 tuples created to our specification. Let's take a look at how `.flatMap()` differs in its implementation. Use the `.flatMap()` method with the same function you created above.

In [26]:
flat_mapped = price_items.flatMap(lambda x: (x,(0.9/1.08)*x))
print(flat_mapped.count())
print(flat_mapped.take(10))

2000
[0.417022004702574, 0.347518337252145, 1.4406489868843162, 1.2005408224035967, 0.0003431244520346599, 0.0002859370433622166, 1.209330290527359, 1.0077752421061326, 0.7337794540855652, 0.6114828784046377]


Rather than being represented by tuples, all of the  values are now on the same level. When we are trying to combine different items together, it is sometimes necessary to use `.flatMap()` rather than `.map()` in order to properly reduce to our specifications. This is not one of those instances, but in the upcoming lab, you just might have to use it.

## Filter
After meeting with some external consultants, BuyStuff has determined that its business will be more profitable if it focuses on higher ticket items. Now, use the `.filter()` method to select items that bring in more than &#36;300 after tax and discount have been removed. A filter method is a specialized form of a map function that only returns the items that match a certain criterion. In the cell below:

* use a lambda function within a `.filter()` method to meet the consultant's suggestion's specifications. set `RDD = selected_items`
* calculate the total number of items remaining in BuyStuff's inventory

In [27]:
# use the filter function
selected_items = discounted.filter(lambda x: x>300)

# calculate total remaining in inventory 
selected_items.count()

272

## Reduce

Reduce functions are where you are in some way combing all of the variables that you have mapped out. Here is an example of how a reduce function works when the task is to sum all values:

<img src = "./images/reduce_function.png" width = "600">  


As you can see, the operation is performed within each partition first, after which, the results of the computations in each partition are combined to come up with one final answer.  

Now it's time to figure out how much money BuyStuff would make from selling one of all of its items after they've reduced their inventory. Use the `.reduce()` method with a lambda function to add up all of the values in the RDD. Your lambda function should have two variables. 

In [28]:
selected_items.reduce(lambda x, y: x + y)

123432.4660522227

The time has come for BuyStuff to open up shop and start selling its goods. It only has one of each item, but it's allowing 50 lucky users to buy as many items as they want while they remain in stock. Within seconds, BuyStuff is sold out. Below, you'll find the sales data in an RDD with tuples of (user, item bought).

In [29]:
# Run this cell without changes
import random
random.seed(42)
# generating simulated users that have bought each item
sales_data = selected_items.map(lambda x: (random.randint(1, 50), x))

sales_data.take(7)

[(45, 303.9717333600381),
 (24, 315.8957533626659),
 (3, 320.6109689548023),
 (12, 320.58739183867544),
 (50, 329.4565433918027),
 (43, 352.7343621192842),
 (49, 336.5077558595325)]

It's time to determine some basic statistics about BuyStuff users.

Let's start off by creating an RDD that determines how much each user spent in total.
To do this we can use a method called `.reduceByKey()` to perform reducing operations while grouping by keys. After you have calculated the total, use the `.sortBy()` method on the RDD to rank the users from the highest spending to the least spending. 

In [30]:
# calculate how much each user spent
total_spent = sales_data.reduceByKey(lambda x, y: x + y)
total_spent.take(10)

[(10, 2717.027724745902),
 (30, 2419.9573922588797),
 (20, 2426.295559078993),
 (50, 1190.2082250357703),
 (40, 3893.7493309486817),
 (1, 3689.0448551795594),
 (31, 4054.2620806396244),
 (41, 2284.03897697932),
 (11, 2661.2893038683637),
 (21, 737.66773326942)]

In [31]:
# sort the users from highest to lowest spenders
total_spent.sortBy(lambda x: x[1], ascending = False).collect()

[(42, 5666.8442832531355),
 (44, 4275.171690198358),
 (34, 4157.22158271992),
 (31, 4054.2620806396244),
 (40, 3893.7493309486817),
 (46, 3872.773968092891),
 (1, 3689.0448551795594),
 (16, 3613.685194600662),
 (47, 3499.406886871411),
 (19, 3279.578676101958),
 (25, 3188.3420049020197),
 (32, 3137.168779592652),
 (13, 3133.856297284863),
 (8, 3126.0614377952056),
 (6, 3027.572032648686),
 (24, 3017.8856513902097),
 (17, 2879.659886470619),
 (5, 2825.2352458866217),
 (43, 2733.935539511269),
 (10, 2717.027724745902),
 (11, 2661.2893038683637),
 (26, 2426.4616983628575),
 (20, 2426.295559078993),
 (30, 2419.9573922588797),
 (27, 2375.494341036703),
 (22, 2366.4128002453563),
 (41, 2284.03897697932),
 (2, 2262.1791827500547),
 (28, 2250.4915604073294),
 (23, 2181.7309320197305),
 (45, 2172.854971180423),
 (37, 2111.532625269436),
 (12, 2107.3695481687037),
 (14, 2017.9570010931986),
 (4, 1966.015875199837),
 (38, 1818.8461805513548),
 (49, 1797.7076367016684),
 (48, 1695.9300524251391),


Next, let's determine how many items were bought per user. This can be solved in one line using an RDD method. After you've counted the total number of items bought per person, sort the users from most number of items bought to least number of items. Time to start a customer loyalty program!

In [32]:
sorted(sales_data.countByKey().items(), key=lambda kv:kv[1], reverse=True)

[(3, 11),
 (34, 10),
 (48, 9),
 (11, 9),
 (16, 8),
 (17, 8),
 (21, 8),
 (38, 8),
 (42, 8),
 (19, 8),
 (33, 8),
 (6, 8),
 (47, 7),
 (50, 7),
 (9, 7),
 (23, 7),
 (13, 6),
 (31, 6),
 (40, 6),
 (49, 6),
 (20, 6),
 (8, 6),
 (2, 5),
 (24, 5),
 (18, 5),
 (35, 5),
 (37, 5),
 (7, 5),
 (4, 4),
 (32, 4),
 (5, 4),
 (28, 4),
 (1, 4),
 (10, 4),
 (26, 4),
 (15, 4),
 (12, 4),
 (22, 4),
 (41, 4),
 (46, 4),
 (29, 4),
 (45, 4),
 (27, 3),
 (36, 3),
 (30, 3),
 (39, 3),
 (25, 2),
 (44, 2),
 (14, 2),
 (43, 1)]

### Stop the SparkContext

Now that we are finished with our analysis, stop the `sc`.

In [33]:
sc.stop()

### Additional Reading

- [The original paper on RDDs](https://cs.stanford.edu/~matei/papers/2012/nsdi_spark.pdf)
- [RDDs in Apache Spark](https://data-flair.training/blogs/create-rdds-in-apache-spark/)
- [Programming with RDDs](https://runawayhorse001.github.io/LearningApacheSpark/rdd.html)
- [RDD Transformations and Actions Summary](https://www.analyticsvidhya.com/blog/2016/10/using-pyspark-to-perform-transformations-and-actions-on-rdd/)

## Summary

In this lab we went through a brief introduction to RDD creation from a Python collection, setting a number of logical partitions for an RDD and extracting lineage. We also used transformations and actions to perform calculations across RDDs on a distributed setup. In the next lab, you'll get the chance to apply these transformations on different books to calculate word counts and various statistics.
